package openinstrument.proto;

message VariableLabel {
  required string name = 1;
  optional string value = 2;
}

message Variable {
  required string name = 1;
  repeated VariableLabel label = 2;
  enum ValueType {
    GAUGE = 0;
    COUNTER = 1;
  }
  optional ValueType value_type = 3 [ default = GAUGE ];
}

message StreamMutation {
  enum SampleType {
    // Do not modify the data at all
    NONE = 0;
    // Average out values between each sampling point
    AVERAGE = 1;
    // Get the maximum value between each sampling point
    MAX = 2;
    // Get the minimum value between each sampling point
    MIN = 3;
    // Get the non-negative rate between each sampling point
    RATE = 4;
    // Get the (possibly) negative rate between each sampling point
    RATE_SIGNED = 5;
    // Difference between each point
    DELTA = 6;
  }
  required SampleType sample_type = 1;

  // Stretch or compress the stream so that there is a value every <sample_frequency> ms.
  // Extra values between each sample will be aggregated according to <sample_type>.
  // Gaps in the stream less than <max_gap_interpolate> samples will be filled with interpolated values between the
  // closest real values.
  // Gaps in the stream of more than <max_gap_interpolate> samples will not be filled.
  optional uint32 sample_frequency = 2;
  optional uint32 max_gap_interpolate = 3 [ default = 1 ];
}

message StreamAggregation {
  enum AggregateType {
    AVERAGE = 0;
    MAX = 1;
    MIN = 2;
    MEDIAN = 3;
    SUM = 4;
  }
  required AggregateType type = 1;

  // Labels to aggregate by on the input streams. If no labels are specified, aggregation will be done on the variable
  // only.
  repeated string label = 2;

  // Points will be aggregated if they are less than sample_interval ms apart. Default is 30 seconds.
  optional uint32 sample_interval = 3 [ default = 30000 ];
}

message Value {
  // Milliseconds since epoch
  required uint64 timestamp = 1;
  required double value = 2;
}

message ValueStream {
  // TODO(dparrish): Remove old_variable once all users are done with it
  optional string old_variable = 1;
  optional Variable variable = 5;
  repeated Value value = 4;
}

message GetRequest {
  // TODO(dparrish): Remove old_variable once all users are done with it
  optional string old_variable = 1;
  optional Variable variable = 9;

  // Milliseconds since epoch
  optional uint64 min_timestamp = 2;
  optional uint64 max_timestamp = 3;

  // If mutations or aggregations are supplied, the GetResponse will have a separate stream for each change requested.
  repeated StreamMutation mutation = 6;
  repeated StreamAggregation aggregation = 7;

  // Limit to how many variables can be returned in a standard GetRequest. This can be overridden if required.
  optional uint32 max_variables = 8 [ default = 100 ];
}

message GetResponse {
  required bool success = 1;
  optional string errormessage = 2;

  // Contains a separate stream for every mutation requested in GetRequest, in the same order. If no mutations are
  // requested, the response will contain a single stream of the raw data.
  repeated ValueStream stream = 3;
}

message AddRequest {
  repeated ValueStream stream = 1;
}

message AddResponse {
  required bool success = 1;
  optional string errormessage = 2;
}

message ListRequest {
  // TODO(dparrish): Remove old_prefix once all users are done with it
  optional string old_prefix = 1;
  optional Variable prefix = 3;

  // Limit to how many variables can be returned in a standard ListRequest. This can be overridden if required.
  optional uint32 max_variables = 2 [ default = 100 ];
}

message ListResponse {
  required bool success = 1;
  optional string errormessage = 2;

  // Although this uses the ValueStream for returning a list of variables, the expectation is that the value field is
  // empty.
  repeated ValueStream stream = 3;
}

message StoreFileHeader {
  required uint64 start_timestamp = 1;
  required uint64 end_timestamp = 2;
  // TODO(dparrish): Remove old_variable once all users are done with it
  repeated string old_variable = 3;
  repeated Variable variable = 4;
}

